packages/transaction-preview/src/index.ts

```diff
@@ -0,0 +1,259 @@
+  validateSimTransactions,
+  validateTPInit
+} from './validation'
+import TransactionPreview from './views/Index.svelte'
+import initI18N from './i18n/index.js'
+import simulateTransactions from './simulateTransactions.js'
+
+export * from './types.js'
+
+const approved$ = new Subject<boolean>()
+let options: TransactionPreviewInitOptions
+let app: TransactionPreview
+
+export const setContainerElement = (containerElement: string): void => {
+  if (containerElement) {
+    const error = validateSetContainerEl(containerElement)
+
+    if (error) {
+      throw error
+    }
+  }
+  options.containerElement = containerElement
+}
+
+const destroyApp = () => {
+  app.$destroy()
+}
+
+const simTransactions = (
+  txs: [TransactionObject]
+): Promise<SimPlatformResponse> => {
+  if (txs) {
+    const error = validateSimTransactions(txs)
+
+    if (error) {
+      throw error
+    }
+  }
+  return simulateTransactions(options, txs)
+}
+
+const handleRequireApproval = async (
+  app: TransactionPreview,
+  fullProviderRequest: EIP1193Provider['request'],
+  req: {
+    method: string
+    params?: Array<unknown>
+  }
+) => {
+  const approved = await firstValueFrom(approved$)
+  app.$destroy()
+  if (!approved) {
+    throw new ProviderRpcError({
+      code: ProviderRpcErrorCode.ACCOUNT_ACCESS_REJECTED,
+      message: 'User rejected the transaction'
+    })
+  }
+  fullProviderRequest(req)
+}
+
+export const patchProvider = (
+  walletProvider: PatchedEIP1193Provider
+): PatchedEIP1193Provider => {
+  if (!walletProvider) {
+    throw new Error(
+      `An EIP 1193 wallet provider is required to preform patching and 
+      watch for transactions e.g. an injected wallet using window.ethereum`
+    )
+  }
+  if (walletProvider.simPatched) return walletProvider as PatchedEIP1193Provider
+
+  const fullProviderRequest = walletProvider.request
+  const patchedProvider = walletProvider as PatchedEIP1193Provider
+  const request: EIP1193Provider['request'] = async (req: {
+    method: string
+    params?: Array<unknown>
+  }): Promise<any> => {
+    if (
+      req.method === 'eth_sendTransaction' &&
+      req.params &&
+      req.params.length
+    ) {
+      const transactionParams =
+        req.params as EthSignTransactionRequest['params']
+      try {
+        const preview = await simulateTransactions(options, transactionParams)
+        if (preview.status !== 'simulated') {
+          // If transaction simulation was unsuccessful do not create DOM el
+          return
+        }
+        if (app) app.$destroy()
+        app = mountTransactionPreview(preview)
+        options.requireTransactionApproval
+          ? handleRequireApproval(app, fullProviderRequest, req)
+          : fullProviderRequest(req).then((hash: unknown) => {
+              hash && app.$destroy()
+            }).catch(() => app.$destroy())
+      } catch (e) {
+        throw new ProviderRpcError({
+          code: ProviderRpcErrorCode.ACCOUNT_ACCESS_REJECTED,
```

Comment:
Is this the correct code for transaction rejected?

Reply:
@aaronbarnardsound Thats a good question! When a transaction is rejected, atleast through MetaMask the code is 4001 the same as `ACCOUNT_ACCESS_REJECTED`
I could create a new enum with the same value for this situation within the common package as this is the second time this code in this context has come up and would add clarity. Thoughts?
<img width="689" alt="Screen Shot 2022-11-14 at 3 48 52 PM" src="https://user-images.githubusercontent.com/24457880/201786093-f9aa8f47-fc42-48cf-8338-59c3d0d24c94.png">

Reply:
No worries, just double checking. No need to add to the enum I don't think.